File: main/src/main/java/com/sedmelluq/discord/lavaplayer/track/info/AudioTrackInfoBuilder.java
package com.sedmelluq.discord.lavaplayer.track.info;

import com.sedmelluq.discord.lavaplayer.tools.DataFormatTools;
import com.sedmelluq.discord.lavaplayer.tools.io.SeekableInputStream;
import com.sedmelluq.discord.lavaplayer.track.AudioReference;
import com.sedmelluq.discord.lavaplayer.track.AudioTrackInfo;

/**
 * Builder for {@link AudioTrackInfo}.
 */
public class AudioTrackInfoBuilder implements AudioTrackInfoProvider {
  private static final String UNKNOWN_TITLE = "Unknown title";
  private static final String UNKNOWN_ARTIST = "Unknown artist";

  private String title;
  private String author;
  private Long length;
  private String identifier;
  private String uri;
  private Boolean isStream;

  private AudioTrackInfoBuilder() {

  }

  @Override
  public String getTitle() {
    return title;
  }

  @Override
  public String getAuthor() {
    return author;
  }

  @Override
  public Long getLength() {
    return length;
  }

  @Override
  public String getIdentifier() {
    return identifier;
  }

  @Override
  public String getUri() {
    return uri;
  }

  public AudioTrackInfoBuilder setTitle(String value) {
    title = DataFormatTools.defaultOnNull(value, title);
    return this;
  }

  public AudioTrackInfoBuilder setAuthor(String value) {
    author = DataFormatTools.defaultOnNull(value, author);
    return this;
  }

  public AudioTrackInfoBuilder setLength(Long value) {
    length = DataFormatTools.defaultOnNull(value, length);
    return this;
  }

  public AudioTrackInfoBuilder setIdentifier(String value) {
    identifier = DataFormatTools.defaultOnNull(value, identifier);
    return this;
  }

  public AudioTrackInfoBuilder setUri(String value) {
    uri = DataFormatTools.defaultOnNull(value, uri);
    return this;
  }

  public AudioTrackInfoBuilder setIsStream(Boolean stream) {
    isStream = stream;
    return this;
  }

  /**
   * @param provider The track info provider to apply to the builder.
   * @return this
   */
  public AudioTrackInfoBuilder apply(AudioTrackInfoProvider provider) {
    if (provider == null) {
      return this;
    }

    return setTitle(provider.getTitle())
        .setAuthor(provider.getAuthor())
        .setLength(provider.getLength())
        .setIdentifier(provider.getIdentifier())
        .setUri(provider.getIdentifier());
  }

  /**
   * @return Audio track info instance.
   */
  public AudioTrackInfo build() {
    long finalLength = DataFormatTools.defaultOnNull(length, Long.MAX_VALUE);

    return new AudioTrackInfo(
        title,
        author,
        finalLength,
        identifier,
        DataFormatTools.defaultOnNull(isStream, finalLength == Long.MAX_VALUE),
        uri
    );
  }

  /**
   * Creates an instance of an audio track builder based on an audio reference and a stream.
   *
   * @param reference Audio reference to use as the starting point for the builder.
   * @param stream Stream to get additional data from.
   * @return An instance of the builder with the reference and track info providers from the stream preapplied.
   */
  public static AudioTrackInfoBuilder create(AudioReference reference, SeekableInputStream stream) {
    AudioTrackInfoBuilder builder = new AudioTrackInfoBuilder()
        .setAuthor(UNKNOWN_ARTIST)
        .setTitle(UNKNOWN_TITLE)
        .setLength(Long.MAX_VALUE);

    builder.apply(reference);

    for (AudioTrackInfoProvider provider : stream.getTrackInfoProviders()) {
      builder.apply(provider);
    }

    return builder;
  }

  /**
   * @return Empty instance of audio track builder.
   */
  public static AudioTrackInfoBuilder empty() {
    return new AudioTrackInfoBuilder();
  }
}

File: main/src/main/java/com/sedmelluq/discord/lavaplayer/track/info/AudioTrackInfoBuilder.kt
package com.sedmelluq.discord.lavaplayer.track.info

import com.sedmelluq.discord.lavaplayer.tools.Units.DURATION_MS_UNKNOWN
import com.sedmelluq.discord.lavaplayer.tools.io.SeekableInputStream
import com.sedmelluq.discord.lavaplayer.track.AudioReference
import com.sedmelluq.discord.lavaplayer.track.AudioTrackInfo

/**
 * Builder for [AudioTrackInfo].
 */
class AudioTrackInfoBuilder internal constructor() : AudioTrackInfoProvider {
    companion object {
        private const val UNKNOWN_TITLE = "Unknown title"
        private const val UNKNOWN_ARTIST = "Unknown artist"

        fun apply(build: AudioTrackInfoBuilder.() -> Unit): AudioTrackInfoBuilder {
            return AudioTrackInfoBuilder()
                .apply(build)
        }

        /**
         * Creates an instance of an audio track builder based on an audio reference and a stream.
         *
         * @param reference Audio reference to use as the starting point for the builder.
         * @param stream    Stream to get additional data from.
         * @return An instance of the builder with the reference and track info providers from the stream pre-applied.
         */
        @JvmStatic
        @JvmOverloads
        fun create(
            reference: AudioReference?,
            stream: SeekableInputStream?,
            build: AudioTrackInfoBuilder.() -> Unit = {}
        ) = apply {
            author = UNKNOWN_ARTIST
            title = UNKNOWN_TITLE
            length = DURATION_MS_UNKNOWN

            if (stream != null) {
                for (provider in stream.trackInfoProviders) {
                    apply(provider)
                }
            }

            apply(build)
            apply(reference)
        }

        /**
         * @return Empty instance of audio track builder.
         */
        @JvmStatic
        fun empty(): AudioTrackInfoBuilder {
            return AudioTrackInfoBuilder()
        }
    }

    override var title: String? = null
    override var author: String? = null
    override var length: Long? = null
    override var identifier: String? = null
    override var uri: String? = null
    override var artworkUrl: String? = null

    var isStream: Boolean? = null

//    fun setTitle(value: String?): AudioTrackInfoBuilder {
//        title = DataFormatTools.defaultOnNull(value, title)
//        return this
//    }
//
//    fun setAuthor(value: String?): AudioTrackInfoBuilder {
//        author = DataFormatTools.defaultOnNull(value, author)
//        return this
//    }
//
//    fun setLength(value: Long?): AudioTrackInfoBuilder {
//        length = DataFormatTools.defaultOnNull(value, length)
//        return this
//    }
//
//    fun setIdentifier(value: String?): AudioTrackInfoBuilder {
//        identifier = DataFormatTools.defaultOnNull(value, identifier)
//        return this
//    }
//
//    fun setUri(value: String?): AudioTrackInfoBuilder {
//        uri = DataFormatTools.defaultOnNull(value, uri)
//        return this
//    }
//
//    fun setArtworkUrl(artworkUrl: String?): AudioTrackInfoBuilder {
//        this.artworkUrl = artworkUrl
//        return this
//    }

    fun setIsStream(stream: Boolean?): AudioTrackInfoBuilder {
        isStream = stream
        return this
    }

    /**
     * @param provider The track info provider to apply to the builder.
     * @return this
     */
    fun apply(provider: AudioTrackInfoProvider?): AudioTrackInfoBuilder {
        return if (provider == null) {
            this
        } else apply {
            title = provider.title
            author = provider.author
            length = provider.length
            identifier = provider.identifier
            uri = provider.uri
            artworkUrl = provider.artworkUrl
        }
    }


    /**
     * @return Audio track info instance.
     */
    fun build(): AudioTrackInfo {
        val length = length ?: DURATION_MS_UNKNOWN
        return AudioTrackInfo(
            title!!,
            author!!,
            length,
            identifier!!,
            uri,
            artworkUrl,
            isStream ?: (length == DURATION_MS_UNKNOWN)
        )
    }
}
